### PR TITLE
pixiecore: initial commit

### PR DESCRIPTION
**Pull request by RaunoVV**
###### Motivation for this change
Initial commit for pixiecore tool: https://github.com/danderson/netboot/tree/master/pixiecore

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Lassulus** [comment]
can we have seperate commits for the service and the package? Also the merge commit is kinda unnecessary

File: pkgs/tools/networking/pixiecore/default.nix
Comment (Lassulus) [inline pkgs/tools/networking/pixiecore/default.nix]: you should add yourself to https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix and reference the attribute then here

File: nixos/modules/services/networking/pixiecore.nix
Comment (Lassulus) [inline nixos/modules/services/networking/pixiecore.nix]: why is this here if it's empty?

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: This is implied so you can omit if you wish.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: You're missing 1 space it appears.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: Please use `mkEnableOption` instead.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: Indentation.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: Indentation.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: Indentation.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: Indentation.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: Indentation.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: This is deprecated please remove.

File: pkgs/tools/networking/pixiecore/default.nix
Comment (aanderse) [inline pkgs/tools/networking/pixiecore/default.nix]: Please set to
```
pname = "netboot";
version = "unstable-2019-02-14";
```
Then `name` will be taken care of for you.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: Am I correct in that you were using a user for the systemd service but then switched to root because of port < 1024 issues? Fortunately modern distros like NixOS allow you to run services which can bind to ports < 1024 without requiring root so I suggest you look into that as running services as root is highly discouraged if not necessary.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: Is this logging to the syslog or a file?

**Review by aanderse — CHANGES_REQUESTED** [review]
Please let me know if you have any questions or need any help.

**Comment by RaunoVV** [comment]
Thank you for the feedback. Will resolve all those today and resubmit! Should it be new clean pullrequest, right?

**Comment by Lassulus** [comment]
You can just amend your commits and force-push to the same branch again

**Comment by aanderse** [comment]
Please do a rebase and then you can avoid the merge in your git history as well. Let us know if you need any help.

File: nixos/modules/services/networking/pixiecore.nix
Comment (RaunoVV) [reply nixos/modules/services/networking/pixiecore.nix]: It should log to syslog

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: @RaunoVV I'm starting to wonder if this service can run entirely without a dedicated user but instead take advantage of [DynamicUser](https://www.freedesktop.org/software/systemd/man/systemd.exec.html) instead.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: This is a privileged port so you'll need to use capabilities of you wish to bind to it. I believe you're looking for cap_net_bind_service. Or just use an unprivileged port.

File: nixos/modules/services/networking/pixiecore.nix
Comment (RaunoVV) [reply nixos/modules/services/networking/pixiecore.nix]: Actually thinking same as its just providing some dhcp, webserver and tftp services for pxe booting on network. Will switch to Dynamic one then :)

File: nixos/modules/services/networking/pixiecore.nix
Comment (RaunoVV) [reply nixos/modules/services/networking/pixiecore.nix]: Will Add   AmbientCapabilities = "cap_net_bind_service"; Should keep using 80 as it's pixiecore's own default port.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: If this is logging to syslog then won't time stamps always be added anyways? If this is true will there be 2 timestamps for every entry?

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: Maybe default should be `null`.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: There is a `port` type.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: There is a `port` type.

File: pkgs/tools/networking/pixiecore/default.nix
Comment (aanderse) [inline pkgs/tools/networking/pixiecore/default.nix]: As mentioned `name` is no longer required.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [reply nixos/modules/services/networking/pixiecore.nix]: Might be worth splitting this over multiple lines.

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [inline nixos/modules/services/networking/pixiecore.nix]: Again, maybe `null` is more appropriate default.

**Review by aanderse — CHANGES_REQUESTED** [review]
Please rewrite your git history.

File: nixos/modules/services/networking/pixiecore.nix
Comment (Ekleog) [reply nixos/modules/services/networking/pixiecore.nix]: Looks like this option is never actually used, so I'd vote for just deleting the `mkOption` block.

File: nixos/modules/services/networking/pixiecore.nix
Comment (Ekleog) [reply nixos/modules/services/networking/pixiecore.nix]: Having it be `types.string` (without `nullOr` and `default = ""` would make more sense to me, unless no command line should be different from an empty command line.

(Also note that as currently implemented the `--cmdline` argument is passed iff `cmdLine` is not `""`, so it'll be passed if it's `null` and this will probably break)

File: nixos/modules/services/networking/pixiecore.nix
Comment (aanderse) [reply nixos/modules/services/networking/pixiecore.nix]: Good catch.

File: nixos/modules/services/networking/pixiecore.nix
Comment (Ekleog) [inline nixos/modules/services/networking/pixiecore.nix]:
```suggestion
        ExecStart  = "${pkgs.pixiecore}/bin/pixiecore boot ${cfg.kernel} ${cfg.initrd} --cmdline='${cfg.cmdLine}' ${optionalString cfg.debug "--debug"} ${optionalString cfg.logTimestamps "--log-timestamps"} ${optionalString cfg.dhcpNoBind "--dhcp-no-bind"} --listen-addr '${cfg.listen}' --port ${toString cfg.port} --status-port ${toString cfg.statusPort}";
```

Unless `--cmdline ""` is different from no `--cmdline` argument. (Also note that as currently written I think you have one too many escapes here, which would not get triggered when staying at the default `--cmdline` though)

Similarly, `cfg.listen` already defaults to `0.0.0.0`, so it can just always be passed. And same for `cfg.port`/`statusPort`, as they're already set to the `pixiecore` default, they can just be always passed. (not sure whether there's necessarily a status port, though? in which case it should be of type `nullOr port` and not be passed when `null` -- not when `0`).

Sure it's hardcoding the defaults of `pixiecore` into the module, but it sounds better to me to just have the whole thing simpler.

Oh, and it's a really long line even with the suggested changes, so you may want to break it over multiple lines -- for instance by closing the NixOS string and opening it again.

File: pkgs/top-level/all-packages.nix
Comment (Ekleog) [inline pkgs/top-level/all-packages.nix]: This is not the right place, it should go under the `### TOOLS` heading (a good trick is to grep for `tools/networking` and place it close to a match)

**Review by Ekleog — CHANGES_REQUESTED** [review]
@RaunoVV Courage! You're almost there, we know it's a lot of feedback to handle… but sorry, you started directly with a package+module, so you get all the feedback you'd normally get over the course of several PRs spread over weeks or months :p So courage! The PR should soon be ready :)

**Comment by aanderse** [comment]
@RaunoVV do you have any motivation to continue with this PR?

**Comment by RaunoVV** [comment]
@aanderse thx for the mention. I'll try to finish it this week. Totally forgot to finish this..

**Comment by ivan** [comment]
For the author, reviewers, and committers: this PR was scanned and appears to add a use of the deprecated `types.string`, which emits a warning as of #66346. Before merging, please change this to another type, possibly:

- `types.str` for a single string where merging does not make sense, or cannot work
- `types.lines` for multi-line configuration or scripts where merging is possible
- `types.listOf types.str` for a mergeable list of strings

**Comment by bbigras** [comment]
I'm looking forward for this module. Any progress on it?

**Comment by Mic92** [comment]
added in https://github.com/NixOS/nixpkgs/pull/83373

**Pull request closed without merging**
